zerver/lib/actions.py

```diff
@@ -2273,27 +2274,33 @@ def notify_subscriptions_removed(user_profile, streams, no_log=False):
 def bulk_remove_subscriptions(users, streams, acting_user=None):
     # type: (Iterable[UserProfile], Iterable[Stream], Optional[UserProfile]) -> Tuple[List[Tuple[UserProfile, Stream]], List[Tuple[UserProfile, Stream]]]
 
-    recipients_map = bulk_get_recipients(Recipient.STREAM,
-                                         [stream.id for stream in streams])  # type: Mapping[int, Recipient]
-    stream_map = {}  # type: Dict[int, Stream]
-    for stream in streams:
-        stream_map[recipients_map[stream.id].id] = stream
+    users = list(users)
+    streams = list(streams)
 
-    subs_by_user = dict((user_profile.id, []) for user_profile in users)  # type: Dict[int, List[Subscription]]
-    for sub in Subscription.objects.select_related("user_profile").filter(user_profile__in=users,
-                                                                          recipient__in=list(recipients_map.values()),
-                                                                          active=True):
-        subs_by_user[sub.user_profile_id].append(sub)
+    stream_ids = {stream.id for stream in streams}
+    stream_dict = {stream.id: stream for stream in streams}
+
+    info = get_bulk_stream_subscriber_info(users, streams)
 
     subs_to_deactivate = []  # type: List[Tuple[Subscription, Stream]]
     not_subscribed = []  # type: List[Tuple[UserProfile, Stream]]
     for user_profile in users:
-        recipients_to_unsub = set([recipient.id for recipient in recipients_map.values()])
-        for sub in subs_by_user[user_profile.id]:
-            recipients_to_unsub.remove(sub.recipient_id)
-            subs_to_deactivate.append((sub, stream_map[sub.recipient_id]))
-        for recipient_id in recipients_to_unsub:
-            not_subscribed.append((user_profile, stream_map[recipient_id]))
+        user_info = info[user_profile.id]
+
+        subcribed_stream_ids = {
+            item['stream'].id
+            for item in user_info
+        }
+        not_subscribed_stream_ids = stream_ids - subcribed_stream_ids
+
+        for stream_id in not_subscribed_stream_ids:
+            stream = stream_dict[stream_id]
+            not_subscribed.append((user_profile, stream))
+
+        for item in user_info:
+            sub = item['sub']
+            stream = item['stream']
+            subs_to_deactivate.append((sub, stream))
```

Comment:
I kinda want more descriptive variables for `info` and `user_info`.  If we were to move the data structure to be a `(sub, stream)` tuple, I guess we could do just do `sub_tuple` and `user_sub_tuples`, or even just `for (sub, stream) in user_sub_tuples`.  Do you think that'd be more readable?

Or if we don't like the tuple idea, `sub_dict` and `user_sub_dicts` might be reasonable.